### PR TITLE
[1/2] No ambiguous field access — Update tests & hack answer tests

### PR DESCRIPTION
**Pull request by cphyc**
Attempt to autofix all possibly ambiguous field accesses.

This replaces *any* occurrence (even the unambiguous ones) of `ad["field_name"]` by `ad[("guessed_field_type", "field_name")]`, where `"guessed_field_type"` is the field type yt used to guess prior to @matthewturk PR #2967 .

The aim of this PR is to fix all issues so that #2967 becomes very simple to address.

### CODE REVIEW

**Comment by neutrinoceros** [comment]
Before I jump in and try to review this beast, do you know if we should worry about example snippet that's embedded in the documentation ? I think Jenkins runs them all so you've probably fixed them already.

**Comment by cphyc** [comment]
I won't be able to work on that in the coming week, so here is a recap' of where this PR stands:
- [ ] the answer tests all passed last time I checked. However, they fail as the answers names have changed from `Projection_density` to `Projection_(gas, density)`. We should either revert to the old behaviour or update the tests.
- [ ] all unit tests pass except one (`yt.frontends.ytdata.tests.test_unit:test_plot_data`). Any help on this one would be appreciated! See notes below.
- [ ] tests which require datasets haven't been updated _yet_, though they should be easier to update than the unit tests.
- [ ] the last task which may be quite tricky is to update the code snippet used to document the classes and functions all over the place. I haven't started to touch them yet, but that means that the doc is going to fail once ambiguous access aren't allowed. I would suggest to postpone that to another PR though.
- [ ] Whenever you run tests on github CI, a diff trying to fix all ambiguous field access is uploaded as an artifact. The same happens with jenkins (see the end of the log). The diff can the be applied to the code base to magically solve the failing line(s). See the second note.

## Notes on `yt.frontends.ytdata.tests.test_unit:test_plot_data`

As far as I understand, the projection/slice operations store the information about the projected location of each cell in the dataset (in the fields `px,py,pz`). When the container is later stored on-disk, these projected values are stored along. When we then reload the dataset, the projected values are reloaded and lead to the creation of `("gas", "px")` and `("all", "px")` (which is rather confusing). @matthewturk correctly points out that these fields are then ambiguous and fails.

I am unsure what best approach should be used.

The old behaviour was to randomly load either `("gas", "px")` or `("all", "px")` and use it to regenerate plots. But that means that I don't know what this piece of code does (and I haven't time to check, but I suspect this leads to the new projection axis to be ignored?)
```python
ds = fake_random_ds(16)
plot = SlicePlot(ds, "z", ("gas", "density"))
fn = plot.data_source.save_as_dataset("slice.h5")
ds_slice = yt.load(fn)
p = SlicePlot(ds_slice, "x", ("gas", "density"))  # note the new projection axis here!
```

## Notes on automatic fixing
This PR includes @matthewturk PR (#2967).
I also added some code that tries to magically replace any field access which is not in the form `(ftype, fname)` by its `(ftype, fname)` inferred type. For example, the following will be replaced as:
```python
ad["density"]  # will become
ad[("gas", "density")]
```
To do so, I catch any field access that is not using a tuple (I hacked `yt/data_objects/static_output.py` for this), walk the stack up until I find the line that accessed the data, load the file where the line is defined and replace it with the equivalent tuple. This replaces the file **in-place** and potentially introduces bugs.

If you want to help and use this evil machinery, you should probably comment out @matthewturk's code that raises an exception (https://github.com/yt-project/yt/pull/2967/files#diff-40060fdda354ceafd86f1ebae4bf9993dc7257656c14dc57822d03c2275d7250R821), otherwise most lines that should have been updated will instead raise an exception. So you either get the exception or the automatic fixing, but unfortunately you can't have both at the same time.

As of 2a9eefc, no exception is raised when an ambiguous field is being accessed, and the automatic-fixing machinery is activated. On jenkins, I hacked `nose_runner.py` so that after all tests have been done, the diff is uploaded using `yt upload` (see the log to find the URL to the diff).

**Comment by neutrinoceros** [comment]
just a random thought: maybe blacken-docs could be hacked in a way that would let us run your auto-fixer in documentation files ?

**Comment by munkm** [comment]
> to please Jenkins


Is it ever possible to truly please Jenkins?

File: yt/geometry/coordinates/cartesian_coordinates.py [hunk 1]
```diff
@@ -534,10 +534,10 @@ def _ortho_pixelize(
...
             pixelize_cartesian(
                 buff,
-                data_source["px"],
-                data_source["py"],
-                data_source["pdx"],
-                data_source["pdy"],
+                data_source[("all", "px")],
+                data_source[("all", "py")],
+                data_source[("all", "pdx")],
+                data_source[("all", "pdy")],
```
Comment (cphyc) [inline yt/geometry/coordinates/cartesian_coordinates.py:540]: Really unsure about these ones. I'd need some input from @matthewturk  on how these `px` and `pdx` thinggies work.

Note that my doubts extend in the entire file.

**Comment by cphyc** [comment]
Currently answer tests are failing because the name of the answer changed (they now include the ftype in addition to the fname).

**Comment by cphyc** [comment]
@matthewturk At the moment almost all answer tests fail due to the answer tests keys being changed to use tuple field name access. Do you have an idea of how to fix this?

Here is my poor attempt at doing this 50dc4c8fd094ff18b5269b281781e527e20affbe (but we should get rid of this though).

File: yt/utilities/exceptions.py [hunk 2]
```diff
@@ -91,8 +91,24 @@ def __init__(self, field, ds):
         self.field = field
         self.ds = ds
 
+    def _enumerate_alternatives(self):
```
Comment (cphyc) [inline yt/utilities/exceptions.py:94]: Note for reviewers: this should be part of #2967 if anything.

**Comment by cphyc** [comment]
**State of this PR**
Checked boxes are activated, unchecked are deactivated
- [x] Ambiguous field detection is activated and raises an error (implementing #2967)
- [x] Non-existing fields raise an exception with suggestions of possible `(ftype, fname)` that have the same `fname` but a different `ftype`
- [ ] The testing infrastructure autofixes the code at locations where field access is ambiguous and report them as build artifacts

**TODO:**
- [x] Update unit tests to remove ambiguous field access
- [x] Update answer tests to remove ambiguous field access
- [x] Update docstring code snippets to remove ambiguous field access
- [x] Update cookbook to remove ambiguous field access
- [x] All unit tests pass, except one (I think we should fix it in another PR)
- [ ] All answer tests pass

File: yt/data_objects/data_containers.py [hunk 3]
```diff
@@ -462,25 +470,24 @@ def write_out(self, filename, fields=None, format="%0.16e"):
...
         for field in field_order:
-            self[field]
+            field_shapes[field] = self[field].shape
+
+        # Check all fields have the same shape
+        all_field_shapes = {shape for shape in field_shapes.values()}
+        if len(all_field_shapes) != 1:
+            err_msg = ["Got fields with different number of elements:\n"]
+            for shape in all_field_shapes:
+                these_fields = [field for field, val in field_shapes.items()]
```
Comment (cphyc) [inline yt/data_objects/data_containers.py:487]: It seems the error message isn't quite correct, try for example:
```python
ds = yt.testing.fake_random_ds(16, particles=10)
sp = ds.all_data()
sp.write_out("/tmp/foo.txt", fields=[("all", "particle_ones")])
```
which returns
```
YTException: Got fields with different number of elements:

         [('all', 'particle_ones'), ('index', 'dx'), ('index', 'dy'), ('index', 'dz'), ('index', 'x'), ('index', 'y'), ('index', 'z')] with shape (4096,)
         [('all', 'particle_ones'), ('index', 'dx'), ('index', 'dy'), ('index', 'dz'), ('index', 'x'), ('index', 'y'), ('index', 'z')] with shape (10,)
```
```suggestion
                these_fields = [
                    field for field, s in field_shapes.items() if s == shape
                ]
```

File: yt/data_objects/tests/test_spheres.py [hunk 4]
```diff
@@ -110,29 +113,29 @@ def test_sphere_center():
...
 
     # Check that any smaller sphere is missing some particles
     sp2 = ds.sphere(sp1.center, sp1.radius * 0.9)
-    N2 = len(sp2["particle_ones"])
+    N2 = len(sp2[("all", "particle_ones")])
     assert N2 < N0
 
 
-@requires_module("MiniballCpp")
+@requires_module("miniball")
```
Comment (cphyc) [inline yt/data_objects/tests/test_spheres.py:135]: This was causing the test to be skipped and never actually tested!

File: yt/fields/tests/test_ambiguous_fields.py [hunk 5]
```diff
@@ -0,0 +1,10 @@
+# from yt.testing import assert_equal, assert_raises, fake_particle_ds
```
Comment (cphyc) [inline yt/fields/tests/test_ambiguous_fields.py:1]: Should be part of #2967 .

File: yt/utilities/answer_testing/framework.py [hunk 6]
```diff
@@ -368,6 +370,18 @@ def sim_dir_load(sim_fn, path=None, sim_type="Enzo", find_outputs=False):
     )
 
 
+def lax_getter(key, dict_like):
```
Comment (cphyc) [inline yt/utilities/answer_testing/framework.py:373]: This should be removed before merging (used to handle mismatch in answer tests keys)

File: yt/utilities/answer_testing/framework.py [hunk 7]
```diff
@@ -421,14 +435,28 @@ def __call__(self):
...
             # Compare test generated values against the golden answer
             dd = self.reference_storage.get(self.storage_name)
-            if dd is None or self.description not in dd:
-                raise YTNoOldAnswer(f"{self.storage_name} : {self.description}")
-            ov = dd[self.description]
+            desc_tried = []
+            if dd is None:
+                raise YTNoOldAnswer(f"{self.storage_name} does not exist.")
+
+            for desc in self.brute_force_description:
```
Comment (cphyc) [inline yt/utilities/answer_testing/framework.py:442]: Should also disappear.

File: yt/utilities/answer_testing/framework.py [hunk 8]
```diff
@@ -480,12 +508,57 @@ def description(self):
...
         args = [self._type_name, str(self.ds), oname]
-        args += [str(getattr(self, an)) for an in self._attrs]
+        # If a field enters the answer test name, we only
+        # keep the field name and drop the field type, i.e.
+        # ("gas", "density") -> "density"
+        for an in self._attrs:
+            tmp = getattr(self, an)
+            if "field" in an and isinstance(tmp, tuple) and len(tmp) == 2:
+                tmp = tmp[1]
+            args.append(str(tmp))
```
Comment (cphyc) [inline yt/utilities/answer_testing/framework.py:518]: Should be reverted before merging.

File: yt/utilities/answer_testing/framework.py [hunk 9]
```diff
@@ -480,12 +508,57 @@ def description(self):
...
         suffix = getattr(self, "suffix", None)
         if suffix:
             args.append(suffix)
         return "_".join(args).replace(".", "_")
 
+    @property
+    def brute_force_description(self):
```
Comment (cphyc) [inline yt/utilities/answer_testing/framework.py:525]: Should be removed before merging.

**Review by cphyc — COMMENTED** [review]
I _think_ this should be ready to be reviewed now! I left many comments in the PR to guide the review.

File: yt/data_objects/tests/test_data_containers.py [hunk 10]
```diff
@@ -61,9 +61,10 @@ def test_yt_data_container(self):
 
     def test_write_out(self):
         filename = "sphere.txt"
-        ds = fake_random_ds(16)
+        ds = fake_random_ds(16, particles=10)
```
Comment (neutrinoceros) [inline yt/data_objects/tests/test_data_containers.py:64]: curious: why did you add this kwarg ?

**Review by neutrinoceros — COMMENTED** [review]
here's a first pass. I found one or two lines that I'm not sure about.
I'm at 57/189 files viewed.

File: yt/data_objects/level_sets/clump_tools.py [hunk 11]
```diff
@@ -70,7 +70,7 @@ def clump_list_sort(clump_list):
     """Returns a copy of clump_list, sorted by ascending minimum density.  This
     eliminates overlap when passing to
     yt.visualization.plot_modification.ClumpContourCallback"""
-    minDensity = [c["Density"].min() for c in clump_list]
+    minDensity = [c[("gas", "density")].min() for c in clump_list]
```
[the remark this answers is not in the document]
Comment (cphyc) [reply yt/data_objects/level_sets/clump_tools.py:73]: I had the impression that fields starting with a capital letter were internal to each dataset (e.g. `("ramses", "Density")`, `("enzo", "Density")`, but then some datasets use `("whatever", "dens")` instead, etc.
So here I have changed to used `("gas", "density")` instead, which should be a common ground to all datasets having a fluid.

[the remark this answers is not in the document]
Comment (neutrinoceros) [reply yt/data_objects/level_sets/clump_tools.py:73]: makes sense. Thanks !

File: yt/data_objects/tests/test_data_containers.py (hunk 10 above)
Comment (cphyc) [reply yt/data_objects/tests/test_data_containers.py:64]: It is to check that writing out the dataset to a file works even if the dataset contains particles and mesh data. This is the complement of `test_invalid_write_out` which tests that writing the dataset *doesn't* work if you try to dump both mesh and particle data at the same time.

Comment (neutrinoceros) [reply yt/data_objects/tests/test_data_containers.py:64]: excellent, thanks

File: yt/data_objects/data_containers.py [hunk 12]
```diff
@@ -154,6 +154,12 @@ def apply_units(self, arr, units):
         except AttributeError:
             return self.ds.arr(arr, units=units)
 
+    def _first_matching_field(self, field):
+        # FIXME? Should be moved elsewhere?
```
[the remark this answers is not in the document]
Comment (cphyc) [reply yt/data_objects/data_containers.py:158]: Yes, and this should raise an exception if no field is found.

**Review by cphyc — COMMENTED** [review]
Now 
- [x] all unit tests pass (except 2 that I deactivated and should be fixed in another PR)
- [x] all answer tests pass 
- [x] except 4 tests for the ART frontend (https://tests.yt-project.org/job/yt_py38_git/2714/testReport/yt.frontends.art.tests/)
- [x] except 7 tests for the ytdata frontend (https://tests.yt-project.org/job/yt_py38_git/2714/testReport/yt.frontends.ytdata.tests/test_outputs/)

At this point, I don't understand why they do not work and since I don't have the answer tests locally I cannot easily debug this. Any help is welcome @matthewturk @Xarthisius @brittonsmith !

Assuming the latter two points are addressed and we are this PR didn't break anything, the next steps are:

1. Update the (nose) answer tests so that they consistently use field tuples in their keys + remove all the current hacks in the code that are used to match the keys in the answer tests
2. Remove the ambiguous field detection (should be in #2967) + reactivate the two deactivated unit tests
3. Merge the PR
4. Re-introduce the ambiguous field detection (it will probably fail due to the two aforementioned unit tests, but that's another PR's job).

**Comment by neutrinoceros** [comment]
We definitely should merge this ASAP, it took you long enough and I expect it to rot extremely quick if we let it rest.
I'll try to complete my review tonight (or this weekend)

File: yt/data_objects/data_containers.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (cphyc) [reply yt/data_objects/data_containers.py]: Done with a test, see f398e88.

File: yt/data_objects/tests/test_projection.py [hunk 13]
```diff
@@ -55,33 +55,44 @@ def test_projection(pf):
...
+                for proj_field in [
+                    ("index", "ones"),
+                    ("gas", "density"),
+                    ("gas", "temperature"),
+                ]:
                     fi = ds._get_field_info(proj_field)
                     assert_equal(frb[proj_field].info["data_source"], proj.__str__())
                     assert_equal(frb[proj_field].info["axis"], ax)
-                    assert_equal(frb[proj_field].info["field"], proj_field)
+                    assert_equal(frb[proj_field].info["field"], str(proj_field))
```
Comment (neutrinoceros) [inline yt/data_objects/tests/test_projection.py:95]: why is this additional conversion necessary ?

**Review by neutrinoceros — COMMENTED** [review]
Ok I'm sorry this is taking so long (75 out of 188)
I naively tried to review this in two steps:
- check as many files as possible which only have additional ptypes/ftypes
- review the rest (infrastructural stuff)

As I'm going forward it looks like there's a larger and larger chance that I'm not going to be able to do that in a reasonable amount of time. I'm really sorry, I know you spent a enormous amount of time and energy into this, but would it be possible to split the infrastructure into a separate PR ? In any case I'll complete this eventually.

**Comment by cphyc** [comment]
> energy

I don't think I can manage to split it properly, the issue being that the infra changes are required for the tests to pass. I could potentially  extract some changes (for example `yt/data_objects/selection_objects/data_selection_objects.py`) into another PR, but tbh I don't really have the stamina or time to do it anytime soon.

I can however suggest *not to review* `framework.py` at all. It is full of temporary hacks to have the answer tests pass and will be dumped the second all answer tests are checked and corrected for.

**Comment by neutrinoceros** [comment]
Fair enough, thank you for this quick reply. I hope I can get done with this in time for our next triage meeting.

File: yt/data_objects/tests/test_projection.py (hunk 13 above)
Comment (neutrinoceros) [reply yt/data_objects/tests/test_projection.py:95]: cool, thanks !

**Comment by cphyc** [comment]
For reference, here is a (hopefully complete) list of files containing changes that aren't simply replacing a non-tuple field access by a tuple field access (though they may contain such changes):
- `yt/data_objects/data_containers.py`
- `yt/data_objects/selection_objects/data_selection_objects.py`
- `yt/data_objects/static_output.py` and `yt/fields/field_info_container.py` and yt/fields/tests/test_ambiguous_fields.py (part of #2967, you can ignore them safely as I'll revert the changes before merging the PR)
- `yt/utilities/answer_testing/framework.py` should be ignored altogether, as it only includes hacks to make the answer tests pass. I will remove all of them before getting this PR in.
- `yt/utilities/exceptions.py` (part of it is part of #2967, I'll remove it)
- `yt/utilities/grid_data_format/writer.py`
- `yt/visualization/profile_plotter.py`

As a guide, all `test_*.py` files contain only changes of the form `field` → `(ftype, field)`.

File: yt/utilities/parallel_tools/io_runner.py [hunk 14]
```diff
@@ -173,7 +173,7 @@ def io_nodes(fn, n_io, n_work, func, *args, **kwargs):
 
     def gq(ds):
         dd = ds.all_data()
-        return dd.quantities["TotalQuantity"]("CellMassMsun")
+        return dd.quantities["TotalQuantity"](("gas", "cell_mass"))
```
Comment (neutrinoceros) [inline yt/utilities/parallel_tools/io_runner.py:176]: wow this one isn't trivial to me. Any insight ?

File: yt/frontends/halo_catalog/tests/test_outputs.py [hunk 15]
```diff
@@ -88,7 +84,7 @@ def test_halo_catalog_boundary_particles(self):
         for field in ["particle_mass"]:
```
Comment (neutrinoceros) [inline yt/frontends/halo_catalog/tests/test_outputs.py:84]: looks like this test is a bit broken
```suggestion
        for field in fields:
```

**Review by neutrinoceros — COMMENTED** [review]
167/188: I cheated and selectively checked the smallest diffs (my browser won't load the whole thing anyway). I'll be back.

File: yt/utilities/parallel_tools/io_runner.py (hunk 14 above)
Comment (cphyc) [reply yt/utilities/parallel_tools/io_runner.py:176]: As far as I could tell, `CellMassMsun` is mapped to `cell_mass` but I may be wrong. Either way, this is an example hidden in the `__main__` section of the file.

@matt I see you are one of the few authors of the file, what is its purpose? I don't see any reference in the code or the doc to it.

File: yt/data_objects/selection_objects/data_selection_objects.py [hunk 16]
```diff
@@ -692,6 +692,24 @@ def cut_region(self, field_cuts, field_parameters=None, locals=None):
...
+            )
+        return self.cut_region(field_cuts)
+
+    def _function_helper(self, function, field, units=None, **kwargs):
+        ftype, fname = self._determine_fields(field)[0]
+        if units is None:
+            field_cuts = f'{function}(obj["{ftype}", "{fname}"])'
+        else:
+            field_cuts = f'{function}(obj["{ftype}", "{fname}"].in_units("{units}"))'
+        return self.cut_region(field_cuts, **kwargs)
```
[the remark this answers is not in the document]
Comment (cphyc) [reply yt/data_objects/selection_objects/data_selection_objects.py]: I improved the name (hopefully) and added a docstring. I hope that's better now!

File: yt/frontends/chombo/tests/test_outputs.py [hunk 17]
```diff
@@ -7,9 +7,9 @@
 )
 
 _fields = (
-    "density",
-    "velocity_magnitude",  # "velocity_divergence",
-    "magnetic_field_x",
+    ("gas", "density"),
+    ("gas", "velocity_magnitude"),  # ("gas", "velocity_divergence"),
```
[the remark this answers is not in the document]
Comment (cphyc) [reply yt/frontends/chombo/tests/test_outputs.py]: Done!

File: yt/data_objects/selection_objects/data_selection_objects.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (neutrinoceros) [reply yt/data_objects/selection_objects/data_selection_objects.py]: much better, thanks !

File: yt/data_objects/data_containers.py [hunk 18]
```diff
@@ -174,11 +181,13 @@ def _set_center(self, center):
             elif center.lower() in ("max", "m"):
                 self.center = self.ds.find_max(("gas", "density"))[1]
             elif center.startswith("max_"):
-                self.center = self.ds.find_max(center[4:])[1]
+                field = self._first_matching_field(center[4:])
```
Comment (neutrinoceros) [inline yt/data_objects/data_containers.py:184]: It's good that you're making the de facto behaviour explicit here, could it be added to a docstring ?

Comment (cphyc) [reply yt/data_objects/data_containers.py:185]: I would rather add that in a new PR, since the `sp = ds.sphere("maximum_density", [...])` syntax doesn't seem to even be in the doc.

Comment (cphyc) [reply yt/data_objects/data_containers.py:185]: I have created a PR to keep track of this, as the function isn't even documented (see #3185).

File: yt/data_objects/data_containers.py [hunk 19]
```diff
@@ -462,25 +471,26 @@ def write_out(self, filename, fields=None, format="%0.16e"):
...
+
+        # Check all fields have the same shape
+        all_field_shapes = {shape for shape in field_shapes.values()}
+        if len(all_field_shapes) != 1:
+            err_msg = ["Got fields with different number of elements:\n"]
+            for shape in all_field_shapes:
+                these_fields = [
+                    field for field, s in field_shapes.items() if s == shape
+                ]
+                err_msg.append(f"\t {these_fields} with shape {shape}")
```
[the remark this answers is not in the document]
Comment (cphyc) [reply yt/data_objects/data_containers.py]: Should be good now!

**Comment by cphyc** [comment]
I manually checked that ART dataset and the answers are correct. The errors we were seeing are because the old answers contain duplicate entries (`density` and `("gas", "density")`), so the assertion in https://github.com/cphyc/yt/blob/d0ebd8727f157b911ddefbc0f4edf7971b54e4f2/yt/utilities/answer_testing/framework.py#L753-L753 fails.

I am bumping the answer to fix it.

File: tests/tests.yaml [hunk 20]
```diff
@@ -1,6 +1,6 @@
 answer_tests:
 
-  local_art_002: # PR 3081
+  local_art_003: # PR 3081, 3101
```
Comment (cphyc) [inline tests/tests.yaml:3]: Required as the old answer tests had duplicate entries (with a different format). This was causing the tests to fail as there are no duplicate entries any more.

**Comment by cphyc** [comment]
The docs build seem unrelated. I think it is ready for final reviews. Once you're happy, I will...
1. remove the part that tests ambiguous fields
2. update answer tests
3. have it merged?

**Comment by Xarthisius** [comment]
@yt-fido test this please.

File: yt/data_objects/data_containers.py (hunk 18 above)
Comment (neutrinoceros) [reply yt/data_objects/data_containers.py:185]: fine with me

**Review by neutrinoceros — COMMENTED** [review]
186/187
The one file I still need to review (yt/utilities/answer_testing/framework.py) requires much higher brain activity, I'll finish this up later !

File: yt/utilities/answer_testing/framework.py [hunk 21]
```diff
@@ -15,6 +15,7 @@
 import urllib
```
Comment (cphyc) [inline yt/utilities/answer_testing/framework.py:15]: For reviewers: this file will entirely be reverted to the version on main. With the exception of lines 466 and 1085-1086, it only contains hacks to make the new answers match the old answers. These should disappear when bumping the answer tests.

**Review by cphyc — COMMENTED** [review]
For reviewers, you can ignore safely ` yt/utilities/answer_testing/framework.py` in your review!

**Review by neutrinoceros — APPROVED** [review]
> For reviewers, you can ignore safely yt/utilities/answer_testing/framework.py in your review!

I forgot about that. In that case here's my formal approval !

File: tests/tests.yaml (hunk 20 above)
Comment (matthewturk) [reply tests/tests.yaml:3]: fascinating

File: yt/visualization/plot_modifications.py [hunk 22]
```diff
@@ -1960,11 +1962,11 @@ def __call__(self, plot):
         px, py = self._convert_to_plot(plot, [px, py])
 
         # Convert halo radii to a radius in pixels
-        radius = halo_data[self.radius_field][:].in_units(units)
+        radius = halo_data[("all", self.radius_field)][:].in_units(units)
```
Comment (matthewturk) [inline yt/visualization/plot_modifications.py:1965]: I'm not sure this is right.  @brittonsmith what do you think?  Is there a specific halo type that we should use for the `HaloCatalog` particle types?

File: yt/visualization/streamlines.py [hunk 23]
```diff
@@ -229,7 +229,7 @@ def path(self, streamline_id):
         >>> streamlines = Streamlines(ds, [0.5]*3)
         >>> streamlines.integrate_through_volume()
         >>> stream = streamlines.path(0)
-        >>> matplotlib.pylab.semilogy(stream['t'], stream['Density'], '-x')
```
Comment (matthewturk) [inline yt/visualization/streamlines.py:232]: funnily enough, I think this would *not* be ambiguous!

**Review by matthewturk — APPROVED** [review]
I love this.  Thank you for your really hard work on this.

File: yt/data_objects/selection_objects/data_selection_objects.py [hunk 24]
```diff
@@ -692,6 +692,48 @@ def cut_region(self, field_cuts, field_parameters=None, locals=None):
         )
         return cr
 
+    def _build_operator_cut(self, operation, field, value, units=None):
```
[the remark this answers is not in the document]
Comment (cphyc) [reply yt/data_objects/selection_objects/data_selection_objects.py:695]: The reason it is here is that the previous syntax was not allowing tuple fields. Take for example `include_equal`, it used to read:
```python
    if units is None:
        field_cuts = 'obj["' + field + '"] == ' + str(value)
```
this used to throw a `TypeError: can only concatenate str (not "tuple") to str`. So I started handling tuple fields, realised there was _a lot_ of code duplication, and decided to only fix this issue once and for all!

File: yt/utilities/exceptions.py (hunk 2 above)
Comment (cphyc) [reply yt/utilities/exceptions.py:94]: Actually should be part of https://github.com/yt-project/yt/pull/3184. I'm removing it.

File: yt/visualization/plot_modifications.py [hunk 25]
```diff
@@ -405,6 +406,7 @@ def __init__(
         self.plot_args = plot_args
 
     def __call__(self, plot):
+        ftype = plot.data._current_fluid_type
```
[the remark this answers is not in the document]
Comment (cphyc) [reply yt/visualization/plot_modifications.py:409]: At the moment what would be the guessed `ftype` in your contrived case?

File: yt/visualization/streamlines.py (hunk 23 above)
Comment (cphyc) [reply yt/visualization/streamlines.py:232]: I wouldn't, but I went ahead and fixed all occurrences of `Density` altogether, as it may be named something else in other frontends.

File: yt/data_objects/data_containers.py [hunk 26]
```diff
@@ -154,6 +155,13 @@ def apply_units(self, arr, units):
         except AttributeError:
             return self.ds.arr(arr, units=units)
 
+    def _first_matching_field(self, field):
```
[the remark this answers is not in the document]
Comment (cphyc) [reply yt/data_objects/data_containers.py:158]: I've opened an issue about this to prevent further delays to this PR, if this is fine for you (#3199 for the reference).

File: yt/data_objects/data_containers.py (hunk 18 above)
Comment (cphyc) [reply yt/data_objects/data_containers.py:185]: Yes! We may want to implement some simple ways of accessing field tuples only by their field name (either first-matching or discarding on-disks).

**Review by cphyc — COMMENTED** [review]
Thanks for the comments @matthewturk!

File: yt/utilities/answer_testing/framework.py (hunk 6 above)
Comment (munkm) [reply yt/utilities/answer_testing/framework.py:404]: So if this is going to get merged in the triage meeting today should it be deleted now?

File: yt/visualization/plot_modifications.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (munkm) [reply yt/visualization/plot_modifications.py:409]: @matthewturk do you have any alternative suggestions for what to do here?

File: yt/visualization/profile_plotter.py [hunk 27]
```diff
@@ -456,9 +456,9 @@ def from_profiles(cls, profiles, labels=None, plot_specs=None, y_log=None):
         >>> plot_specs = []
         >>> for ds in es[-4:]:
         ...     ad = ds.all_data()
-        ...     profiles.append(create_profile(ad, ["Density"],
-        ...                                    fields=["Temperature",
-        ...                                            "x-velocity"]))
+        ...     profiles.append(create_profile(ad, [("gas, "density")],
+        ...                                    fields=[("gas", "temperature"),
+        ...                                            ("gas", "velocity_x")]))
```
Comment (munkm) [inline yt/visualization/profile_plotter.py:461]: good catch on the velocity here.

**Review by munkm — APPROVED** [review]
This looks great! I have a few questions here but I am happy for this to go in as is.

File: yt/data_objects/selection_objects/cut_region.py [hunk 28]
```diff
@@ -27,15 +27,15 @@ class YTCutRegion(YTSelectionContainer3D):
         A list of conditionals that will be evaluated.  In the namespace
         available, these conditionals will have access to 'obj' which is a data
         object of unknown shape, and they must generate a boolean array.  For
-        instance, conditionals = ["obj['temperature'] < 1e3"]
+        instance, conditionals = ["obj['gas', 'temperature'] < 1e3"]
```
[the remark this answers is not in the document]
Comment (munkm) [reply yt/data_objects/selection_objects/cut_region.py:30]: Oh I figured out why! Anyways, this doesn't need to be updated in this PR but I think it would be good later on to make these tuples so we're consistent in the codebase.

File: yt/visualization/plot_modifications.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (matthewturk) [reply yt/visualization/plot_modifications.py:409]: Nope, I think it's fine.

File: yt/data_objects/selection_objects/cut_region.py (hunk 28 above)
[the remark this answers is not in the document]
Comment (cphyc) [reply yt/data_objects/selection_objects/cut_region.py:30]: I _tried_ to use consistently the latter format, but I obviously missed a few of them. I think the ideal outcome would be to have that enforced by `black` :)

File: yt/data_objects/tests/test_extract_regions.py [hunk 29]
```diff
@@ -29,50 +29,54 @@ def test_cut_region():
         dd = ds.all_data()
         r = dd.cut_region(
             [
-                "obj['temperature'] > 0.5",
-                "obj['density'] < 0.75",
-                "obj['velocity_x'] > 0.25",
+                "obj['gas', 'temperature'] > 0.5",
+                "obj['gas', 'density'] < 0.75",
+                "obj['gas', 'velocity_x'] > 0.25",
```
[the remark this answers is not in the document]
Comment (cphyc) [reply yt/data_objects/tests/test_extract_regions.py:34]: ⇒ https://github.com/yt-project/yt/issues/3201

File: yt/data_objects/selection_objects/cut_region.py (hunk 28 above)
[the remark this answers is not in the document]
Comment (neutrinoceros) [reply yt/data_objects/selection_objects/cut_region.py:30]: I don't think black can do that, for reference, because in the general case, a two-element key isn't equivalent to a single two-size tuple one.

File: yt/frontends/halo_catalog/tests/test_outputs.py [hunk 30]
```diff
@@ -85,10 +81,10 @@ def test_halo_catalog_boundary_particles(self):
 
         assert type(ds) is YTHaloCatalogDataset
 
-        for field in ["particle_mass"]:
+        for field in fields:
             f1 = data[field].in_base()
```
[the remark this answers is not in the document]
Comment (cphyc) [reply yt/frontends/halo_catalog/tests/test_outputs.py:85]: No, `data` is defined as
```
        data = {
            field: YTArray(rs.random_sample(n_halos), unit)
            for field, unit in zip(fields, units)
        }
```
so it should be indexed by items in `fields`, whatever they are

[the remark this answers is not in the document]
Comment (munkm) [reply yt/frontends/halo_catalog/tests/test_outputs.py:85]: 👍

File: yt/frontends/halo_catalog/tests/test_outputs.py [hunk 31]
```diff
@@ -56,17 +54,15 @@ def test_halo_catalog(self):
...
             assert_array_equal(f1, f2)
 
     @requires_module("h5py")
     def test_halo_catalog_boundary_particles(self):
         rs = np.random.RandomState(3670474)
         n_halos = 100
-        fields = [
-            f"particle_{name}" for name in ["mass"] + [f"position_{ax}" for ax in "xyz"]
-        ]
+        fields = ["particle_mass"] + [f"particle_position_{ax}" for ax in "xyz"]
```
[the remark this answers is not in the document]
Comment (cphyc) [reply yt/frontends/halo_catalog/tests/test_outputs.py:65]: I don't know to be honest. But since the tests used to pass, this isn't strictly necessary.

File: yt/frontends/chombo/fields.py [hunk 32]
```diff
@@ -257,15 +265,15 @@ def setup_particle_fields(self, ptype, ftype="gas", num_neighbors=64):
...
 
 
 def _dummy_velocity(field, data):
-    return np.zeros_like(data["particle_velocity_x"])
+    return np.zeros_like(data[("all", "particle_velocity_x")])
 
 
 def _dummy_field(field, data):
-    return 0.0 * data["gravitational_field_x"]
+    return 0.0 * data[("chombo", "gravitational_field_x")]
```
[the remark this answers is not in the document]
Comment (cphyc) [reply yt/frontends/chombo/fields.py:276]: Is this even tested anywhere? I'm surprised it doesn't cause any issue.

File: yt/utilities/answer_testing/framework.py (hunk 6 above)
Comment (cphyc) [reply yt/utilities/answer_testing/framework.py:404]: No, it is deleted in #3194 by commit 7771d30750474b404e7ca40ad25bd3d4f917e79d!

File: yt/data_objects/selection_objects/cut_region.py (hunk 28 above)
[the remark this answers is not in the document]
Comment (cphyc) [reply yt/data_objects/selection_objects/cut_region.py:30]: Agreed!

File: yt/frontends/chombo/fields.py (hunk 32 above)
[the remark this answers is not in the document]
Comment (cphyc) [reply yt/frontends/chombo/fields.py:276]: According to https://github.com/cphyc/yt/blob/d8654b73050dbc6b023c986ea67ba337faed9148/yt/frontends/chombo/fields.py#L335-L335, it seems to be defined with an underscore? Unless I am mistaken
